core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -170,13 +170,29 @@ private[spark] class DAGScheduler(
    */
   private val cacheLocs = new HashMap[Int, IndexedSeq[Seq[TaskLocation]]]
 
-  // For tracking failed nodes, we use the MapOutputTracker's epoch number, which is sent with
-  // every task. When we detect a node failing, we note the current epoch number and failed
-  // executor, increment it for new tasks, and use this to ignore stray ShuffleMapTask results.
+  // Tracks the latest epoch of a fully processed error related to the given executor. (We use
+  // the MapOutputTracker's epoch number, which is sent with every task.)
   //
-  // TODO: Garbage collect information about failure epochs when we know there are no more
-  //       stray messages to detect.
-  private val failedEpoch = new HashMap[String, Long]
+  // When an executor fails, it can affect the results of many tasks, and we have to deal with
+  // all of them consistently. We don't simply ignore all future results from that executor,
+  // as the failures may have been transient; but we also don't want to "overreact" to follow-
+  // on errors we receive. Furthermore, we might receive notification of a task success, after
+  // we find out the executor has actually failed; we'll assume those successes are, in fact,
+  // simply delayed notifications and the results have been lost, if they come from the same
+  // epoch. In particular, we use this to control when we tell the BlockManagerMaster that the
+  // BlockManager has been lost.
```

Comment:
As tasks started for an earlier epoch are also considered to be lost:

https://github.com/apache/spark/blob/d09ef9335e5d3657b830497155abb7a0c2bb0cde/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1585-L1586

Please mention that case too, like:

```suggestion
  // simply delayed notifications and the results have been lost, if those tasks started for an
  // earlier or the same epoch. In particular, we use this to control when we tell the 
  // BlockManagerMaster that the BlockManager has been lost.
```

Reply:
Adopted.